src/main/java/core/basesyntax/service/RegistrationServiceImpl.java

```
@@ -2,13 +2,44 @@
 
 import core.basesyntax.dao.StorageDao;
 import core.basesyntax.dao.StorageDaoImpl;
+import core.basesyntax.exception.NotValidUserException;
 import core.basesyntax.model.User;
 
 public class RegistrationServiceImpl implements RegistrationService {
+    private static final int MIN_AGE = 18;
+    private static final int MIN_LENGTH = 6;
     private final StorageDao storageDao = new StorageDaoImpl();
 
     @Override
     public User register(User user) {
-        return null;
+        if (user == null) {
+            throw new NotValidUserException("User can not be null");
+        }
+        if (user.getLogin() == null) {
+            throw new NotValidUserException("Login should not be null");
+        }
+        User existing = storageDao.get(user.getLogin());
+        if (existing != null) {
+            throw new NotValidUserException("User is already registered");
+        }
+        if (user.getAge() == null) {
+            throw new NotValidUserException("User age should not be null");
+        }
+        if (user.getAge() < MIN_AGE) {
+            throw new NotValidUserException("Not a valid age: "
+                    + user.getAge()
+                    + ". Min allowed age is "
+                    + MIN_AGE);
+        }
+        if (user.getLogin().length() < MIN_LENGTH) {
+            throw new NotValidUserException("Login should be at least 6 characters long");
+        }
```

This validation is correct. For better readability and to fail faster on login-related issues, it's a good practice to group all login validations together. Consider moving this check to after the existing user validation (around line 24).